### PR TITLE
Add stdlib reef

### PR DESCRIPTION
**Pull request by Override-6**
This pull request adds a standard library reef and fixes some bugs with function called from foreign reefs.

- The analyzer currently only accepts native function declarations that comes from reef id `1`

### CODE REVIEW

File: parser/src/aspects/function_declaration.rs [hunk 1]
```diff
@@ -32,33 +32,57 @@ impl<'a> FunctionDeclarationAspect<'a> for Parser<'a> {
...
+                return_type: rtype,
+                body: None,
+                segment: segment_start..self.cursor.relative_pos_ctx(token).end,
+            });
+        }
 
         let body = self
             .cursor
-            .force(blanks().then(of_type(Equal)), "expected '='")
+            .force(blanks().then(of_type(Equal)), "`=` expected")
```
Comment (syldium) [inline parser/src/aspects/function_declaration.rs:71]:
```suggestion
            .force(blanks().then(of_type(Equal)), "expected '='")
```
You expect something, not someone.

File: lib/std.msh [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/// makes the runtime panic, with specified message and callstack trace.
```
Comment (syldium) [inline lib/std.msh:1]:
```suggestion
/// Causes the runtime system to panic, with the specified message.
```
The call stack is not a parameter.

File: lib/std.msh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+/// once executed the process will exit with code 1.
+/// @param message the cause of the panic
+fun panic(message: String) -> Nothing;
+
+/// exits program with given exitcode
```
Comment (syldium) [inline lib/std.msh:6]:
```suggestion
/// Exits the current process with the given exitcode.
```

File: lib/std.msh [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+/// @param name the targeted environment variable name.
+/// @param value the new value of targeted variable.
+fun set_env(name: String, value: String);
+
+/// tests if given environment variable is defined
+/// @returns true if the given environment variable has a defined value
+/// @param name the targeted environment variable name
+fun is_env_def(name: String) -> Bool;
+
+/// reads a line from stdin
```
Comment (syldium) [inline lib/std.msh:24]:
```suggestion
/// Reads a single line from the standard input.
///
/// If the input is closed, an empty string is returned.
```

File: lib/std/convert.msh [hunk 5]
```diff
@@ -0,0 +1,12 @@
+/// converts given int value to an exitcode
+/// @panics if i cannot be converted to an exitcode, i.e it is negative or superior to 256
+fun to_exitcode(i: Int) -> Exitcode;
```
Comment (syldium) [inline lib/std/convert.msh:3]: This could be a method on the type `Int`.

File: cli/src/report.rs [hunk 6]
```diff
@@ -88,24 +91,33 @@ pub fn display_diagnostic<W: Write>(
...
+        let engine = if engine_reef == loc.reef {
+            current_engine
+        } else {
+            &externals.get_reef(loc.reef).unwrap().engine
+        };
+
         let content_id = engine
-            .get_original_content(obs.location.source)
+            .get_original_content(loc.source)
             .expect("Unknown source");
```
Comment (syldium) [inline cli/src/report.rs:110]: Panics when importing `std` outside the root directory of this repository:
```bash
cd cli && cargo run <<< 'use std::env'
```

File: analyzer/tests/collect_debug.rs [hunk 7]
```diff
@@ -1,6 +1,7 @@
 use std::collections::HashSet;
 
 use analyzer::analyze;
+
```
Comment (syldium) [inline analyzer/tests/collect_debug.rs:4]:
```suggestion
```

File: analyzer/src/types/engine.rs [hunk 8]
```diff
@@ -66,20 +67,29 @@ impl Chunk {
...
         Self {
-            expression,
+            expression: Some(expression),
             parameters: Vec::new(),
             return_type: UNIT,
         }
     }
 
+    /// Creates a new script chunk.
+    pub fn native_function(parameters: Vec<Parameter>, return_type: TypeRef) -> Self {
```
Comment (syldium) [inline analyzer/src/types/engine.rs:77]:
```suggestion
    pub fn native(parameters: Vec<Parameter>, return_type: TypeRef) -> Self {
```

File: analyzer/src/steps/collect.rs [hunk 9]
```diff
@@ -200,7 +200,10 @@ impl<'a, 'b, 'e> SymbolCollector<'a, 'b, 'e> {
                     );
                     let diagnostic = {
                         Diagnostic::new(DiagnosticID::SymbolConflictsWithModule, msg)
-                            .with_observation(Observation::here(env_id, declaration_segment.clone(), format!("This symbol has the same fully-qualified name as module {}", clashed_module.fqn)))
+                            .with_observation(
+                                Observation::from((env_id, self.externals.current, declaration_segment.clone()))
+                                    .with_message(format!("This symbol has the same fully-qualified name as module {}", clashed_module.fqn))
```
Comment (syldium) [inline analyzer/src/steps/collect.rs:205]:
```suggestion
                                Observation::here(env_id, self.externals.current, declaration_segment.clone(), format!("This symbol has the same fully-qualified name as module {}", clashed_module.fqn))
```

File: analyzer/src/lib.rs [hunk 10]
```diff
@@ -25,6 +25,7 @@ use crate::imports::Imports;
 use crate::name::Name;
 use crate::reef::Externals;
 use crate::relations::{Relations, SourceId};
+
```
Comment (syldium) [inline analyzer/src/lib.rs:28]:
```suggestion
```

**Review by syldium — CHANGES_REQUESTED** [review]

File: cli/src/cli.rs [hunk 11]
```diff
@@ -76,16 +77,18 @@ impl SourceLineProvider for CachedSourceLocationLineProvider {
...
+pub fn use_pipeline(
     entry_point: &Name,
     starting_page: SourceId,
     analyzer: &Analyzer<'_>,
+    externals: &Externals,
     vm: &mut VM,
     diagnostics: Vec<Diagnostic>,
-    importer: &mut (impl ASTImporter<'a> + ErrorReporter),
+    errors: Vec<FileImportError>,
+    sources: &SourcesCache,
```
Comment (syldium) [inline cli/src/cli.rs:89]: Pass the importer directly, as it already contains those two things.

File: cli/src/main.rs [hunk 12]
```diff
@@ -39,40 +42,63 @@ fn main() -> Result<PipelineStatus, miette::Error> {
...
+        path
+    };
+
+    let mut importer = FileImporter::new(sources.len(), folder_path);
+
+    importer.add_redirection(name.clone(), source.to_path_buf());
+
+    let mut analyzer = Analyzer::new();
+
+    analyzer.process(name.clone(), &mut importer, &externals);
```
Comment (syldium) [inline cli/src/main.rs:87]:
```suggestion
    let mut importer = FileImporter::new(sources.len(), folder_path);
    importer.add_redirection(name.clone(), source.to_path_buf());

    let mut analyzer = Analyzer::new();
    analyzer.process(name.clone(), &mut importer, &externals);
```
They are closely related, don't separate them :cry:

File: cli/src/pipeline.rs [hunk 13]
```diff
@@ -86,6 +72,9 @@ pub struct FileImporter {
     /// The root directory from which the files are read.
     root: PathBuf,
 
+    /// a shift for local content ids
+    sources_shift: usize,
+
```
Comment (syldium) [inline cli/src/pipeline.rs:77]:
```suggestion
```
The `SourceCache` is the importer itself.

File: cli/src/pipeline.rs [hunk 14]
```diff
@@ -99,21 +88,37 @@ pub struct FileImporter {
...
+impl SourcesCache {
+    pub fn extend(&mut self, sources: Vec<OwnedSource>) {
+        self.cache.extend(sources)
+    }
+
+    pub fn len(&self) -> usize {
+        self.cache.len()
+    }
+}
+
```
Comment (syldium) [inline cli/src/pipeline.rs:105]:
```suggestion
```
The `SourceCache` is the importer itself.

File: cli/src/pipeline.rs [hunk 15]
```diff
@@ -99,21 +88,37 @@ pub struct FileImporter {
...
 impl FileImporter {
     /// Creates a new file importer that will import sources from the given
     /// root directory.
-    pub fn new(root: PathBuf) -> Self {
+    pub fn new(sources_shift: usize, root: PathBuf) -> Self {
```
Comment (syldium) [inline cli/src/pipeline.rs:109]:
```suggestion
    pub fn new(root: PathBuf) -> Self {
```
A simple constructor would just start at `0`. To ensure `SourceId`s are different, just use the same importer and change the `root` `PathBuf` for instance.

File: cli/src/pipeline.rs [hunk 16]
```diff
@@ -99,21 +88,37 @@ pub struct FileImporter {
...
-            sources: Vec::new(),
             redirections: HashMap::new(),
             errors: Vec::new(),
         }
     }
 
     /// Inserts a new source into the importer.
-    pub fn insert<'a>(&mut self, source: OwnedSource) -> ImportResult<'a> {
-        let id = self.sources.len();
+    pub fn insert<'b>(&mut self, source: OwnedSource) -> ImportResult<'b> {
```
Comment (syldium) [inline cli/src/pipeline.rs:120]: May I ask to keep the `'a`?

File: compiler/src/emit/invoke.rs [hunk 17]
```diff
@@ -134,10 +134,18 @@ pub fn emit_function_invocation(
...
-            let captures = ctx.captures[id.0]
-                .as_ref()
-                .expect("captures not set during function invocation emission");
-            let env = ctx.engine.get_environment(id).unwrap();
+            let captures: &[ResolvedSymbol] = if function_call.reef != ctx.current_reef {
+                &[]
+            } else {
+                ctx.captures[id.0]
+                    .as_ref()
+                    .expect("captures not set during function invocation emission")
```
Comment (syldium) [inline compiler/src/emit/invoke.rs:142]:
```suggestion
                    .expect("undefined captures when the function is emitted")
```
*During* is for long time frames in English, and *emission* is a bit too French for me.

File: analyzer/src/types/engine.rs [hunk 18]
```diff
@@ -66,20 +67,29 @@ impl Chunk {
...
     pub fn script(expression: TypedExpr) -> Self {
         Self {
-            expression,
+            expression: Some(expression),
             parameters: Vec::new(),
             return_type: UNIT,
         }
     }
 
+    /// Creates a new script chunk.
```
Comment (syldium) [inline analyzer/src/types/engine.rs:76]:
```suggestion
    /// Creates a new function that is natively defined.
```

File: analyzer/src/steps/typing.rs [hunk 19]
```diff
@@ -105,6 +104,35 @@ impl TypingState {
     }
 }
 
+fn check_nativity(
```
Comment (syldium) [inline analyzer/src/steps/typing.rs:107]:
```suggestion
fn verify_free_function(
```

**Review by syldium — CHANGES_REQUESTED** [review]

File: cli/src/cli.rs (hunk 11 above)
Comment (Override-6) [reply cli/src/cli.rs:89]: no it does not contains `sources` anymore

File: cli/src/std.rs [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+    }
+}
+
+fn find_std() -> PathBuf {
+    if let Ok(path) = std::env::var("MOSHELL_STD") {
+        return PathBuf::from(path);
+    }
+
+    if let Some(proj_dirs) = ProjectDirs::from("", "", "moshell") {
+        return proj_dirs.data_dir().join("lib").to_path_buf();
```
Comment (syldium) [inline cli/src/std.rs:45]:
```suggestion
        let lib = proj_dirs.data_dir().join("lib");
        return if lib.exists() {
            lib
        } else {
            let mut dir = std::env::current_dir().expect("Could not get current directory");
            dir.push("lib");
            dir
        };
```
This should let a new user use `cargo run` in the current directory without any prior any installation.

File: cli/src/std.rs [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+        diagnostics,
+        importer.take_errors(),
+        sources,
+        config,
+    );
+    externals.register(Reef::new("std".to_string(), analyzer));
+
+    if status != PipelineStatus::Success {
+        panic!("std build did not succeed")
+    }
```
Comment (syldium) [inline cli/src/std.rs:36]:
```suggestion
    match status {
        PipelineStatus::Success => {}
        PipelineStatus::IoError => panic!(
            "Unable to find the standard library, check the MOSHELL_STD environment variable"
        ),
        _ => panic!("std build did not succeed"),
    }

File: compiler/src/lib.rs [hunk 22]
```diff
@@ -242,13 +257,21 @@ fn resolve_captures(engine: &Engine, relations: &Relations, compiled_reef: ReefI
 /// compiles chunk's code attribute
 /// the code attribute of a chunk is a special attribute that contains the bytecode instructions and
 /// locals specifications
+///
+/// returns the hir's segments associated with their first instruction,
+/// and the total amount of compiled instructions.
```
Comment (syldium) [inline compiler/src/lib.rs:262]:
```suggestion
/// returns the hir's segments associated with their first instruction.
```

**Review by syldium — APPROVED** [review]

**Pull request merged**